### ISSUE TITLE
training efficiency

### ISSUE THREAD

**Issue by 6-Qingying** [atomistic-machine-learning/schnetpack#244]
Hello！
![efficiency](https://user-images.githubusercontent.com/73232902/97128007-2f2d2680-1776-11eb-863e-a1450877c294.gif)
My training efficiency on both GPU and CPU is very low even if I add 'num_workers=4' to the learning script. Is there any way to increase efficiency？
Thank you for help! :）

**Comment by Stefaanhess** [comment]
Hey @6-Qingying ,
I assume that you want to increase the volatile gpu-util? Are you using an SSD? You could also try to set `pin_memory=True` in the dataloader. This [post](https://discuss.pytorch.org/t/gpu-high-memory-usage-low-gpu-volatile-util/19856) on the pytorch-forum eyplains possible reasons. Let me know, if that solved the problem!
If you want to increase the memory-usage, you could use a bigger batch size.

**Comment by 6-Qingying** [comment]
`import argparse
import os
import torch
import numpy as np
import schnetpack as spk
import schnetpack.train as trn
from schnetpack import AtomsData
from torch.optim import Adam

parser = argparse.ArgumentParser()
parser.add_argument("-db", type=str, help="Database file path")
parser.add_argument("-log", type=str, help="Log directory")
args = parser.parse_args()

logdir = args.log
if not os.path.exists(logdir):
    os.makedirs(logdir)

new_dataset = AtomsData(args.db,
                        available_properties=['energy'])
train, val, test = spk.train_test_split(
    data=new_dataset,
    num_train=43000,
    num_val=15000,
    split_file=os.path.join(logdir, "split.npz"),
)

train_loader = spk.AtomsLoader(train, batch_size=256, shuffle=True, num_workers=2, pin_memory=True)
val_loader = spk.AtomsLoader(val, batch_size=256, pin_memory=True)

schnet = spk.representation.SchNet(
    n_atom_basis=30, n_filters=60, n_gaussians=20, n_interactions=6,cutoff=5., 
    normalize_filter=False, coupled_interactions=False, return_intermediate=False, 
    trainable_gaussians=False, distance_expansion=None, charged_systems=False, 
    cutoff_network=spk.nn.cutoff.CosineCutoff
)

output_U0 = spk.atomistic.ElementalAtomwise(
    n_in=30, property=new_dataset.available_properties[0], elements=frozenset({1, 6, 7, 8, 16}))
model = spk.AtomisticModel(representation=schnet, output_modules=output_U0)


# loss function
def mse_loss(batch, result):
    diff = batch[new_dataset.available_properties[0]] - \
        result[new_dataset.available_properties[0]]
    err_sq = torch.mean(diff ** 2)
    return err_sq


# build optimizer
optimizer = Adam(model.parameters(), lr=1e-3)

loss = trn.build_mse_loss([new_dataset.available_properties[0]])

metrics = [spk.metrics.MeanAbsoluteError(new_dataset.available_properties[0])]
hooks = [
    trn.CSVHook(log_path=logdir, metrics=metrics),
    trn.ReduceLROnPlateauHook(
        optimizer,
        patience=5, factor=0.8, min_lr=1e-6,
        stop_after_min=True
    )
]

trainer = trn.Trainer(
    model_path=logdir,
    model=model,
    hooks=hooks,
    loss_fn=loss,
    optimizer=optimizer,
    train_loader=train_loader,
    validation_loader=val_loader,
)

device = "cuda"
trainer.train(device=device)

results = np.loadtxt(os.path.join(logdir, 'log.csv'),
                     skiprows=1, delimiter=',')

time = results[:, 0]-results[0, 0]
learning_rate = results[:, 1]
train_loss = results[:, 2]
val_loss = results[:, 3]
val_mae = results[:, 4]

print('Final validation MAE:', np.round(val_mae[-1], 2), 'kcal/mol')`

Hello!
This is my traing script. And my data is not on SSD. My training efficiency doesn't change noticeably after setting `pin_memory=True` in the dataloader.  >_<
 I also wonder if there is someway which could speed up the calculation of every step. 
Thank you for your help (and forgive my pour English) !

**Comment by Dankomaister** [comment]
Hi @6-Qingying,
Maybe I can give some suggestions.
First try and set `num_workers` for both the `train_loader` and `val_loader` since both are used during training.
Also what hardware are you running on?
I would increase the `num_workers` to at least 4 per GPU (depending on how many CPU cores you have available).
I see that you are trying to run on two GPUs, maybe try and use only one initially and see if you can efficiently utilize that.
I also agree with @Stefaanhess that is better if the training data is located on fast storage like an SSD or if you are using Linux you can first copy the training data to a `tmpfs` filesystem like `/dev/shm` which is essentiality a RAM disk and read it form there.

**Comment by Stefaanhess** [comment]
I just did a little test, to measure the difference. When I run a training session on the SSD, I get a volatile gpu-util value between 30% and 90%. The same script on a HDD only yields values between 0% and 30% :)

**Comment by 6-Qingying** [comment]
> Hi @6-Qingying,
> Maybe I can give some suggestions.
> First try and set `num_workers` for both the `train_loader` and `val_loader` since both are used during training.
> Also what hardware are you running on?
> I would increase the `num_workers` to at least 4 per GPU (depending on how many CPU cores you have available).
> I see that you are trying to run on two GPUs, maybe try and use only one initially and see if you can efficiently utilize that.
> I also agree with @Stefaanhess that is better if the training data is located on fast storage like an SSD or if you are using Linux you can first copy the training data to a `tmpfs` filesystem like `/dev/shm` which is essentiality a RAM disk and read it form there.

Thanks for help!
I have tried the approach you provided, and it is quite useful! Thank you so much!

**Comment by 6-Qingying** [comment]
> I just did a little test, to measure the difference. When I run a training session on the SSD, I get a volatile gpu-util value between 30% and 90%. The same script on a HDD only yields values between 0% and 30% :)

That explains a lot!  I see. Thanks again for your help! And SSD is unavailable temporarily, maybe one day I would have a try and experience a volatile gpu-util value between 30% and 90%. 👍 :)
